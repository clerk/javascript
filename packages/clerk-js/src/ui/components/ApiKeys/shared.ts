import { useClerk } from '@clerk/shared/react';
import { useState } from 'react';

import { useFetch } from '../../hooks';
import type { Expiration } from './CreateApiKeyForm';

function getTimeLeftInSeconds(expirationOption: Expiration) {
  if (expirationOption === 'never') {
    return;
  }
  const now = new Date();
  const future = new Date(now);
  if (expirationOption === '30d') {
    future.setDate(future.getDate() + 30);
  } else if (expirationOption === '90d') {
    future.setDate(future.getDate() + 90);
  } else {
    throw new Error('TODO: Improve time handling');
  }
  const diffInMs = future.getTime() - now.getTime();
  const diffInSecs = Math.floor(diffInMs / 1000);
  return diffInSecs;
}

export function useApiKeys({ subject, perPage = 5 }: { subject: string; perPage?: number }) {
  const clerk = useClerk();
  const {
    data: apiKeys,
    isLoading,
    revalidate,
  } = useFetch(clerk.getApiKeys, { subject }, undefined, `api-key-source-${subject}`);
  const [page, setPage] = useState(1);
  const [search, setSearch] = useState('');
  const itemsPerPage = perPage;

  const filteredApiKeys = (apiKeys ?? []).filter(key => key.name.toLowerCase().includes(search.toLowerCase()));
  const itemCount = filteredApiKeys.length;
  const pageCount = Math.max(1, Math.ceil(itemCount / itemsPerPage));
  const startingRow = itemCount > 0 ? (page - 1) * itemsPerPage + 1 : 0;
  const endingRow = Math.min(page * itemsPerPage, itemCount);
  const paginatedApiKeys = filteredApiKeys.slice(startingRow - 1, endingRow);

  const handleCreate = async (params: {
    name: string;
    description?: string;
    expiration: Expiration;
    closeFn: () => void;
  }) => {
    await clerk.createApiKey({
      name: params.name,
      creationReason: params.description,
      secondsUntilExpiration: getTimeLeftInSeconds(params.expiration),
    });
    params.closeFn();
    revalidate();
  };

  const revokeApiKey = async (apiKeyID: string) => {
    await clerk.revokeApiKey({ apiKeyID, revocationReason: 'Revoked by user' });
    setPage(1);
    revalidate();
  };

  return {
    apiKeys: paginatedApiKeys,
    isLoading: isLoading ?? false,
    revokeApiKey,
    search,
    setSearch,
    page,
    setPage,
    pageCount,
    itemCount,
    startingRow,
    endingRow,
    handleCreate,
  };
}
