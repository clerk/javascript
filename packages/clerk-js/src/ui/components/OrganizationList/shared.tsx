import type { UserOrganizationInvitationResource } from '@clerk/types';
import type { PropsWithChildren } from 'react';
import { forwardRef } from 'react';

import { Box, Button, Col, descriptors, Flex, Spinner } from '../../customizables';
import { OrganizationPreview, PreviewButton } from '../../elements';
import { ArrowRightIcon } from '../../icons';
import type { ThemableCssProp } from '../../styledSystem';
import { common } from '../../styledSystem';

export const PreviewListItems = (props: PropsWithChildren) => {
  return (
    <Col
      elementDescriptor={descriptors.organizationListPreviewItems}
      sx={t => ({
        maxHeight: `calc(8 * ${t.sizes.$12})`,
        overflowY: 'auto',
        ...common.unstyledScrollbar(t),
      })}
    >
      {props.children}
    </Col>
  );
};

const sharedStyles: ThemableCssProp = t => ({
  height: t.space.$14,
  padding: `${t.space.$2} ${t.space.$8}`,
});

export const PreviewListItem = (
  props: PropsWithChildren<{
    organizationData: UserOrganizationInvitationResource['publicOrganizationData'];
  }>,
) => {
  return (
    <Flex
      align='center'
      gap={2}
      sx={[
        _ => ({
          minHeight: 'unset',
          justifyContent: 'space-between',
        }),
        sharedStyles,
      ]}
      elementDescriptor={descriptors.organizationListPreviewItem}
    >
      <OrganizationPreview
        elementId='organizationList'
        avatarSx={t => ({ width: t.sizes.$10, height: t.sizes.$10 })}
        mainIdentifierSx={t => ({
          fontSize: t.fontSizes.$xl,
          color: t.colors.$colorText,
        })}
        organization={props.organizationData}
      />
      {props.children}
    </Flex>
  );
};

export const PreviewListSpinner = forwardRef<HTMLDivElement>((_, ref) => {
  return (
    <Box
      ref={ref}
      sx={t => ({
        width: '100%',
        height: t.space.$12,
        position: 'relative',
      })}
    >
      <Box
        sx={{
          margin: 'auto',
          position: 'absolute',
          left: '50%',
          top: '50%',
          transform: 'translateY(-50%) translateX(-50%)',
        }}
      >
        <Spinner
          size='md'
          colorScheme='primary'
        />
      </Box>
    </Box>
  );
});

export const PreviewListItemButton = (props: Parameters<typeof Button>[0]) => {
  return (
    <Button
      elementDescriptor={descriptors.organizationListPreviewItemActionButton}
      textVariant='buttonExtraSmallBold'
      variant='outline'
      colorScheme='neutral'
      size='sm'
      {...props}
    />
  );
};

export const OrganizationListPreviewButton = (props: PropsWithChildren<{ onClick: () => void | Promise<void> }>) => {
  return (
    <PreviewButton
      elementDescriptor={descriptors.organizationListPreviewButton}
      sx={[sharedStyles]}
      icon={ArrowRightIcon}
      iconProps={{
        size: 'lg',
      }}
      showIconOnHover={false}
      {...props}
    />
  );
};
