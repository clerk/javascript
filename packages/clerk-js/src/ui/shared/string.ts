const IP_V4_ADDRESS_REGEX =
  /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;

export function isIPV4Address(str: string | undefined | null): boolean {
  return IP_V4_ADDRESS_REGEX.test(str || '');
}

export function titleize(str: string | undefined | null): string {
  const s = str || '';
  return s.charAt(0).toUpperCase() + s.slice(1);
}

export function snakeToCamel(str: string | undefined): string {
  return str ? str.replace(/([-_][a-z])/g, match => match.toUpperCase().replace(/-|_/, '')) : '';
}

export function camelToSnake(str: string | undefined): string {
  return str ? str.replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`) : '';
}
