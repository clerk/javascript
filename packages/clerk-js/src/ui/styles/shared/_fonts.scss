$font-size-base: 1rem; // Assumes the browser default, typically `16px`

$font-size-lg: $font-size-base * 1.25;
$font-size-sm: $font-size-base * 0.875;
$font-size-xs: $font-size-base * 0.75;

$semibold-weight: 600;

.font-semibold {
  font-weight: $semibold-weight;
}
