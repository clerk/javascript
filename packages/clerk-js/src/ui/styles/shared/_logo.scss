.logo {
  display: block;
  width: 100%;

  margin: 0 auto;
  padding: 0;
  text-align: center;
  white-space: nowrap;

  font-size: 3rem;
  line-height: 3rem;
  height: 3rem;

  color: $theme-primary;

  > img {
    height: 100%;
    vertical-align: initial;
  }
}
