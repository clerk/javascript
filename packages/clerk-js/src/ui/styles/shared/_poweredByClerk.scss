.powered-by-clerk-container {
  display: flex;
  justify-content: center;
  color: $themed-font-color;

  span {
    opacity: 0.3;
    font-size: $font-size-xs;
  }

  .powered-by-clerk-logo {
    opacity: 0.2;
    margin-left: 4px;
    margin-top: -5px;
    color: $themed-font-color;
  }
}
