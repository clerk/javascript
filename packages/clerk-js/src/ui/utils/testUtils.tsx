import { AttributeData, Attributes, Clerk, EnvironmentResource, OAuthProviders, SignInData } from '@clerk/types';
import { jest } from '@jest/globals';
import React from 'react';

import { ComponentContext, EnvironmentProvider } from '../contexts';
import { CoreClerkContextWrapper } from '../contexts/CoreClerkContextWrapper';
import { AppearanceProvider } from '../customizables';
import { FlowMetadataProvider } from '../elements';
import { RouteContext } from '../router';
import { InternalThemeProvider } from '../styledSystem';

const initialConfig = {
  frontendApi: '',
  organization: undefined,
  user: null,
  session: null,
  addListener: jest.fn(),
  client: {
    pathRoot: '/client',
    sessions: [],
    signUp: {
      pathRoot: '/client/sign_ups',
      status: null,
      requiredFields: [],
      optionalFields: [],
      missingFields: [],
      unverifiedFields: [],
      verifications: {
        emailAddress: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
          nextAction: '',
          supportedStrategies: [],
        },
        phoneNumber: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
          nextAction: '',
          supportedStrategies: [],
        },
        web3Wallet: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
          nextAction: '',
          supportedStrategies: [],
        },
        externalAccount: {
          pathRoot: '',
          status: null,
          strategy: null,
          nonce: null,
          externalVerificationRedirectURL: null,
          attempts: null,
          expireAt: null,
          error: null,
          verifiedAtClient: null,
        },
      },
      username: null,
      firstName: null,
      lastName: null,
      emailAddress: null,
      phoneNumber: null,
      web3wallet: null,
      hasPassword: false,
      unsafeMetadata: {},
      createdSessionId: null,
      createdUserId: null,
      abandonAt: null,
    },
    signIn: {
      pathRoot: '/client/sign_ins',
      status: null,
      supportedIdentifiers: [],
      supportedFirstFactors: [],
      supportedSecondFactors: [],
      firstFactorVerification: {
        pathRoot: '',
        status: null,
        strategy: null,
        nonce: null,
        externalVerificationRedirectURL: null,
        attempts: null,
        expireAt: null,
        error: null,
        verifiedAtClient: null,
      },
      secondFactorVerification: {
        pathRoot: '',
        status: null,
        strategy: null,
        nonce: null,
        externalVerificationRedirectURL: null,
        attempts: null,
        expireAt: null,
        error: null,
        verifiedAtClient: null,
      },
      identifier: null,
      createdSessionId: null,
      userData: {},
    },
    lastActiveSessionId: null,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
    id: 'client_2GLpcZTwM08TAm7RCDFrfekjbCO',
  },
} as any; // initialize config

const initialEnvironmentResource = {
  userSettings: {
    id: undefined,
    social: {
      oauth_google: {
        enabled: true,
        required: false,
        authenticatable: true,
        strategy: 'oauth_google',
      },
    } as OAuthProviders,
    attributes: {
      email_address: {
        enabled: true,
        required: true,
        name: 'email_address',
        verifications: ['email_code'],
        used_for_first_factor: true,
        first_factors: ['email_code'],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: true,
      } as AttributeData,
      phone_number: {
        enabled: true,
        required: false,
        name: 'phone_number',
        verifications: ['phone_code'],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: true,
        second_factors: ['phone_code'],
        verify_at_sign_up: true,
      } as AttributeData,
      username: {
        enabled: false,
        required: false,
        name: 'username',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      first_name: {
        enabled: true,
        required: true,
        name: 'first_name',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      last_name: {
        enabled: true,
        required: true,
        name: 'last_name',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      password: {
        enabled: true,
        required: true,
        name: 'password',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      web3_wallet: {
        enabled: false,
        required: false,
        name: 'web3_wallet',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      authenticator_app: {
        enabled: false,
        required: false,
        name: 'authenticator_app',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
      backup_code: {
        enabled: false,
        required: false,
        name: 'backup_code',
        verifications: [],
        used_for_first_factor: false,
        first_factors: [],
        used_for_second_factor: false,
        second_factors: [],
        verify_at_sign_up: false,
      } as AttributeData,
    } as Attributes,
    signIn: {
      second_factor: {
        required: false,
      },
    } as SignInData,
    signUp: {
      custom_action_required: false,
      disable_hibp: false,
      progressive: true,
    },
    socialProviderStrategies: ['oauth_google'],
    authenticatableSocialStrategies: ['oauth_google'],
    web3FirstFactors: [],
    enabledFirstFactorIdentifiers: ['email_address'],
    instanceIsPasswordBased: true,
    hasValidAuthFactor: true,
    pathRoot: '',
  },
  authConfig: {
    pathRoot: '',
    singleSessionMode: false,
  },
  displayConfig: {
    pathRoot: '',
    id: 'display_config_2GLnHgZkgHLIp5F3ciLxw1jdRrX',
    instanceEnvironmentType: 'production',
    applicationName: 'ClerkProd-(local)',
    theme: {
      buttons: {
        font_color: '#ffffff',
        font_family: '"Source Sans Pro", sans-serif',
        font_weight: '600',
      },
      general: {
        color: '#6c47ff',
        padding: '1em',
        box_shadow: '0 2px 8px rgba(0, 0, 0, 0.2)',
        font_color: '#151515',
        font_family: '"Source Sans Pro", sans-serif',
        border_radius: '0.5em',
        background_color: '#ffffff',
        label_font_weight: '600',
      },
      accounts: {
        background_color: '#ffffff',
      },
    },
    preferredSignInStrategy: 'password',
    logoUrl: 'https://images.clerk.services/clerk/logo.svg',
    faviconUrl: null,
    backendHost: '',
    homeUrl: '',
    signInUrl: '',
    signUpUrl: '',
    userProfileUrl: '',
    afterSignInUrl: '',
    afterSignUpUrl: '',
    afterSignOutOneUrl: '',
    afterSignOutAllUrl: '',
    afterSwitchSessionUrl: '',
    branded: true,
    supportEmail: '',
    clerkJSVersion: null,
    experimental__forceOauthFirst: false,
  },
  isSingleSession: () => false,
  isProduction: () => true,
  onWindowLocationHost: () => true,
} as any as EnvironmentResource;

const createClerkFixture = () => {
  const listeners: any[] = [];

  const client = {};
  const session = {};
  const user = {};
  const organization = {};
  const resources = { client, session, user, organization };
  const addListener = (listener: any) => {
    listeners.push(listener);
    listener(resources);
  };
  const updateMock = () => {
    // how do we create new state?
    const newState = { ...resources };
    listeners.forEach(listener => listener(newState));
  };

  return { mockedClerk: { ...resources, addListener } as Clerk, updateMock };
};

type FParam = {
  withUsername: () => void;
  withEmailAddress: () => void;
  withGoogleOAuth: () => void;
  withAllOAuth: () => void;
  withPhoneCode: () => void;
};

type ConfigFn = (f: FParam) => void;

export const createFixture = (configFn: ConfigFn) => {
  const config = initialConfig;
  const f = {
    withUsername: () => {
      // config.user_settings.username = {
      //   enabled: true,
      //   required: true,
      // };
    },
    withGoogleOauth: () => {
      config.user_settings.social.google_oauth = {
        enabled: true,
        required: true,
      };
    },
  } as any as FParam;
  if (configFn) {
    configFn(f);
  }

  const { mockedClerk, updateMock } = createClerkFixture();

  const MockClerkProvider = (props: any) => {
    const { children } = props;
    return (
      <CoreClerkContextWrapper clerk={config}>
        <EnvironmentProvider value={initialEnvironmentResource}>
          <RouteContext.Provider
            value={{
              basePath: '',
              startPath: '',
              fullPath: '',
              indexPath: '',
              currentPath: '',
              queryString: '',
              queryParams: {},
              getMatchData: jest.fn(),
              matches: jest.fn(),
              baseNavigate: jest.fn(),
              navigate: jest.fn(),
              resolve: jest.fn(() => ({ href: 'test.host' } as any)),
              refresh: jest.fn(),
              params: {},
            }}
          >
            <AppearanceProvider appearanceKey={'signIn'}>
              <FlowMetadataProvider flow={'signIn'}>
                <InternalThemeProvider>
                  <ComponentContext.Provider value={{ componentName: 'SignIn' }}>{children}</ComponentContext.Provider>
                </InternalThemeProvider>
              </FlowMetadataProvider>
            </AppearanceProvider>
          </RouteContext.Provider>
        </EnvironmentProvider>
      </CoreClerkContextWrapper>
    );
  };

  return { MockClerkProvider, updateMock };
};
