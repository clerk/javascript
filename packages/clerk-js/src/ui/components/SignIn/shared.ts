import { isClerkRuntimeError, isUserLockedError } from '@clerk/shared/error';
import { useClerk } from '@clerk/shared/react';
import { useCallback, useEffect } from 'react';

import { clerkInvalidFAPIResponse } from '../../../core/errors';
import { __internal_WebAuthnAbortService } from '../../../utils/passkeys';
import { useCoreSignIn, useSignInContext } from '../../contexts';
import { useCardState } from '../../elements';
import { useSupportEmail } from '../../hooks/useSupportEmail';
import { handleError } from '../../utils';

function useHandleAuthenticateWithPasskey(onSecondFactor: () => Promise<unknown>) {
  const card = useCardState();
  // @ts-expect-error -- private method for the time being
  const { setActive, __internal_navigateWithError } = useClerk();
  const supportEmail = useSupportEmail();
  const { afterSignInUrl } = useSignInContext();
  const { authenticateWithPasskey } = useCoreSignIn();

  useEffect(() => {
    return () => {
      __internal_WebAuthnAbortService.abort();
    };
  }, []);

  return useCallback(async (...args: Parameters<typeof authenticateWithPasskey>) => {
    try {
      const res = await authenticateWithPasskey(...args);
      switch (res.status) {
        case 'complete':
          return setActive({ session: res.createdSessionId, redirectUrl: afterSignInUrl });
        case 'needs_second_factor':
          return onSecondFactor();
        default:
          return console.error(clerkInvalidFAPIResponse(res.status, supportEmail));
      }
    } catch (err) {
      const { flow } = args[0] || {};

      if (isClerkRuntimeError(err)) {
        // In any case if the call gets aborted we should skip showing an error. This prevents updating the state of unmounted components.
        if (err.code === 'passkey_operation_aborted') {
          return;
        }
        // In case of autofill, if retrieval of credentials is cancelled by the user avoid showing errors as it results to pour UX.
        if (flow === 'autofill' && err.code === 'passkey_retrieval_cancelled') {
          return;
        }
      }

      if (isUserLockedError(err)) {
        return __internal_navigateWithError('..', err.errors[0]);
      }
      handleError(err, [], card.setError);
    }
  }, []);
}

export { useHandleAuthenticateWithPasskey };
