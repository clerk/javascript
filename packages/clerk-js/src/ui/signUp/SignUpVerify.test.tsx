import { noop, render, renderJSON } from '@clerk/shared/testUtils';
import {
  SessionResource,
  SignInStrategyName,
  SignUpResource,
} from '@clerk/types';
import React from 'react';

import {
  SignUpVerifyEmailAddress,
  SignUpVerifyPhoneNumber,
} from './SignUpVerify';

const navigateMock = jest.fn();
const mockSetSession = jest.fn();
const mockAttemptVerification = jest.fn();
const mockStartMagicLinkFlow = jest.fn(() => {
  return Promise.resolve({
    status: 'complete',
    verifications: {
      emailAddress: {
        verifiedFromTheSameClient: jest.fn(() => false),
        verifiedAtClient: '',
      },
    },
  } as any as SignUpResource);
});
let mockFirstFactors: SignInStrategyName[];
let mockDisabledStrategies: string[] = [];

jest.mock('ui/router/RouteContext');

jest.mock('ui/contexts', () => {
  return {
    useEnvironment: jest.fn(() => ({
      displayConfig: {
        applicationName: 'My test app',
        afterSignUpUrl: 'http://test.host',
      },
      authConfig: {
        username: 'on',
        firstName: 'required',
        lastName: 'required',
        password: 'required',
        firstFactors: mockFirstFactors,
        identificationRequirements: [
          ['email_address', 'phone_address', 'oauth_google', 'oauth_facebook'],
        ],
      },
    })),
    useCoreSession: () => {
      return { id: 'sess_id' } as SessionResource;
    },
    withCoreUserGuard: (a: any) => a,
    useCoreClerk: () => {
      return {
        setSession: mockSetSession,
      };
    },
    useSignUpContext: () => {
      return {
        signInUrl: 'http://test.host/sign-in',
        navigateAfterSignUp: jest.fn(),
        disabledStrategies: mockDisabledStrategies,
      };
    },
    useCoreSignUp: () => {
      return {
        emailAddress: 'jdoe@example.com',
        phoneNumber: '+123456789',
        verifications: {
          emailAddress: {
            status: 'unverified',
            verifiedFromTheSameClient: jest.fn(() => false),
          },
          phoneNumber: {
            status: 'unverified',
          },
        },
        attemptVerification: mockAttemptVerification,
        createMagicLinkFlow: () => ({
          startMagicLinkFlow: mockStartMagicLinkFlow,
          cancelMagicLinkFlow: () => {
            return;
          },
        }),
      };
    },
  };
});

jest.mock('ui/hooks', () => ({
  // @ts-ignore
  ...jest.requireActual('ui/hooks'),
  useNavigate: () => {
    return {
      navigate: navigateMock,
    };
  },
}));

describe('<SignUpVerify/>', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('verify email address', () => {
    it('renders the sign up verification form', async () => {
      mockFirstFactors = ['email_code', 'phone_code', 'password'];
      const tree = renderJSON(<SignUpVerifyEmailAddress />);
      expect(tree).toMatchSnapshot();
    });

    it('renders the sign up verification form but prefers email_link if exists', async () => {
      mockFirstFactors = ['email_code', 'phone_code', 'password', 'email_link'];
      const tree = renderJSON(<SignUpVerifyEmailAddress />);
      expect(tree).toMatchSnapshot();
    });

    it('can skip disabled verification strategies', async () => {
      mockFirstFactors = ['email_code', 'phone_code', 'password', 'email_link'];
      mockDisabledStrategies = ['email_link'];
      const { container } = render(<SignUpVerifyEmailAddress />);
      expect(container.querySelector('.cl-otp-input')).not.toBeNull();
    });

    xit(
      'renders the verifiation screen, types the OTP and attempts an email address verification',
      noop,
    );
  });

  describe('verify phone number', () => {
    it('renders the sign up verification form', async () => {
      const tree = renderJSON(<SignUpVerifyPhoneNumber />);
      expect(tree).toMatchSnapshot();
    });

    xit(
      'renders the verifiation screen, types the OTP and attempts a phone number verification',
      noop,
    );
  });
});
