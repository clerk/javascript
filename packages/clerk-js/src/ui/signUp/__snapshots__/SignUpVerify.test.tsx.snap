// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUpVerify/> verify email address renders the sign up verification form 1`] = `
Array [
  <div
    className="cl-auth-form-header"
    style={Object {}}
  >
    <div
      className="cl-auth-form-header-logo"
    >
      <div
        className="cl-logo"
        title="My test app"
      >
        My test app
      </div>
    </div>
  </div>,
  <div
    className="cl-auth-form-body"
  >
    <label
      className="label cl-auth-form-message"
    >
      Enter the 6-digit code sent to
      <br />
      <strong>
        jdoe@example.com
      </strong>
    </label>
    <div
      className="cl-otp-input"
    >
      <div
        className="inputContainer"
        data-test-id="otp-container"
      >
        <input
          aria-label="Enter verification code. Digit 1"
          autoComplete="one-time-code"
          autoFocus={true}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-0"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 2"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-1"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 3"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-2"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 4"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-3"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 5"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-4"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 6"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-5"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
      </div>
      <div
        className="statusContainer"
      >
        <button
          className="button link primary resendCode"
          disabled={false}
          onClick={[Function]}
        >
          Resend Code
        </button>
      </div>
    </div>
  </div>,
]
`;

exports[`<SignUpVerify/> verify email address renders the sign up verification form but prefers email_link if exists 1`] = `
<div
  className="cl-auth-form-body cl-auth-form-body-center cl-verification-page-container"
>
  <div
    className="cl-verification-page-container"
    data-test-id="magicLink-waiting-screen"
  >
    <div
      className="cl-verification-page-icon-container"
    >
      <svg
        data-filename="../shared/assets/icons/mail-open.svg"
      />
    </div>
    <h1
      className="cl-verification-page-header"
    >
      Check your email
    </h1>
    <p
      className="cl-verification-page-text"
    >
      A sign-up link has been sent to
      <br />
      <span
        className="cl-verification-page-identifier"
      >
        jdoe@example.com
      </span>
    </p>
    <p
      className="cl-verification-page-text"
    >
      Click the link in the email to continue signing up, then return to
          this tab.
    </p>
    <button
      className="button link primary disabled cl-verify-page-resend-button"
      disabled={true}
      onClick={[Function]}
    >
      Resend magic link
    </button>
  </div>
</div>
`;

exports[`<SignUpVerify/> verify phone number renders the sign up verification form 1`] = `
Array [
  <div
    className="cl-auth-form-header"
    style={Object {}}
  >
    <div
      className="cl-auth-form-header-logo"
    >
      <div
        className="cl-logo"
        title="My test app"
      >
        My test app
      </div>
    </div>
  </div>,
  <div
    className="cl-auth-form-body"
  >
    <label
      className="label cl-auth-form-message"
    >
      Enter the 6-digit code sent to
      <br />
      <strong>
        +123456789
      </strong>
    </label>
    <div
      className="cl-otp-input"
    >
      <div
        className="inputContainer"
        data-test-id="otp-container"
      >
        <input
          aria-label="Enter verification code. Digit 1"
          autoComplete="one-time-code"
          autoFocus={true}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-0"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 2"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-1"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 3"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-2"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 4"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-3"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 5"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-4"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
        <input
          aria-label="Digit 6"
          autoComplete="one-time-code"
          autoFocus={false}
          className="input codeInput"
          disabled={false}
          inputMode="numeric"
          maxLength={1}
          name="codeInput-5"
          onChange={[Function]}
          onFocus={[Function]}
          onInput={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          type="text"
          value=" "
        />
      </div>
      <div
        className="statusContainer"
      >
        <button
          className="button link primary resendCode"
          disabled={false}
          onClick={[Function]}
        >
          Resend Code
        </button>
      </div>
    </div>
  </div>,
]
`;
